### PR TITLE
Branch: master. Modification for supporting multithread and clear singleton.

### PR DESCRIPTION
**Pull request by lfcarol**
In order to support multiple threads, I add some lock and clear the singletons.

### CODE REVIEW

**Comment by unacao** [comment]
Lint failed

**Comment by unacao** [comment]
Lint failed: FAIL: (06-21 03:57:47.933) //modules/control/controller:controller_interface_cpplint

**Comment by unacao** [comment]
Please also add a gflag to turn on mutex only in multi-threading

**Comment by unacao** [comment]
This commit will have runtime error for scenarios 10088, 10880 and 10896:

Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000000000096d3c4 in apollo::common::VehicleStateProvider::linear_velocity() const ()
[Current thread is 1 (Thread 0x7f026750d3c0 (LWP 185))]
(gdb) #0  0x000000000096d3c4 in apollo::common::VehicleStateProvider::linear_velocity() const ()
#1  0x000000000061a071 in apollo::planning::RuleBasedStopDecider::CheckADCStop(apollo::planning::PathData const&, apollo::planning::ReferenceLineInfo const&, double) ()
#2  0x000000000061b00a in apollo::planning::RuleBasedStopDecider::StopOnSidePass(apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
#3  0x000000000061b1a9 in apollo::planning::RuleBasedStopDecider::Process(apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
#4  0x000000000062f177 in apollo::planning::Decider::Execute(apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
#5  0x00000000005a12e0 in apollo::planning::scenario::lane_follow::LaneFollowStage::PlanOnReferenceLine(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
#6  0x00000000005a1e35 in apollo::planning::scenario::lane_follow::LaneFollowStage::Process(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*) ()
#7  0x00000000005b72c6 in apollo::planning::scenario::Scenario::Process(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*) ()
#8  0x0000000000591d1a in apollo::planning::PublicRoadPlanner::Plan(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*, apollo::planning::ADCTrajectory*) ()
#9  0x0000000000574c47 in apollo::planning::OnLanePlanning::Plan(double, std::vector<apollo::common::TrajectoryPoint, std::allocator<apollo::common::TrajectoryPoint> > const&, apollo::planning::ADCTrajectory*) ()
#10 0x0000000000576f66 in apollo::planning::OnLanePlanning::RunOnce(apollo::planning::LocalView const&, apollo::planning::ADCTrajectory*) ()
#11 0x00000000005537d3 in apollo::planning::PlanningComponent::Proc(std::shared_ptr<apollo::prediction::PredictionObstacles> const&, std::shared_ptr<apollo::canbus::Chassis> const&, std::shared_ptr<apollo::localization::LocalizationEstimate> const&) ()
#12 0x00000000004f9e19 in apollo::simulation::PlanningTask::RunOnce() ()
#13 0x000000000048407b in apollo::simulation::TaskManager::ExecuteNextTaskAndWrite() ()
#14 0x000000000046df39 in apollo::simulation::ScenarioRunner::Run() ()
#15 0x00000000004351bb in main ()

File: modules/control/controller/BUILD [hunk 1]
```diff
@@ -126,6 +129,7 @@ cc_test(
     name = "lon_controller_test",
     size = "small",
     srcs = ["lon_controller_test.cc"],
+    copts = ["-fno-access-control"],
```
[the remark this answers is not in the document]
Comment (lfcarol) [reply modules/control/controller/BUILD:132]: The "LonControllerTest::Init()" uses "LongController::injector_" which is private member.

**Comment by lx18233184051** [comment]
take care of dead lock :)

**Comment by lfcarol** [comment]
> This commit will have runtime error for scenarios 10088, 10880 and 10896:
> 
> Program terminated with signal SIGSEGV, Segmentation fault.
> #0 0x000000000096d3c4 in apollo::common::VehicleStateProvider::linear_velocity() const ()
> [Current thread is 1 (Thread 0x7f026750d3c0 (LWP 185))]
> (gdb) #0 0x000000000096d3c4 in apollo::common::VehicleStateProvider::linear_velocity() const ()
> #1 0x000000000061a071 in apollo::planning::RuleBasedStopDecider::CheckADCStop(apollo::planning::PathData const&, apollo::planning::ReferenceLineInfo const&, double) ()
> #2 0x000000000061b00a in apollo::planning::RuleBasedStopDecider::StopOnSidePass(apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
> #3 0x000000000061b1a9 in apollo::planning::RuleBasedStopDecider::Process(apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
> #4 0x000000000062f177 in apollo::planning::Decider::Execute(apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
> #5 0x00000000005a12e0 in apollo::planning::scenario::lane_follow::LaneFollowStage::PlanOnReferenceLine(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*, apollo::planning::ReferenceLineInfo*) ()
> #6 0x00000000005a1e35 in apollo::planning::scenario::lane_follow::LaneFollowStage::Process(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*) ()
> #7 0x00000000005b72c6 in apollo::planning::scenario::Scenario::Process(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*) ()
> #8 0x0000000000591d1a in apollo::planning::PublicRoadPlanner::Plan(apollo::common::TrajectoryPoint const&, apollo::planning::Frame*, apollo::planning::ADCTrajectory*) ()
> #9 0x0000000000574c47 in apollo::planning::OnLanePlanning::Plan(double, std::vector<apollo::common::TrajectoryPoint, std::allocatorapollo::common::TrajectoryPoint > const&, apollo::planning::ADCTrajectory*) ()
> #10 0x0000000000576f66 in apollo::planning::OnLanePlanning::RunOnce(apollo::planning::LocalView const&, apollo::planning::ADCTrajectory*) ()
> #11 0x00000000005537d3 in apollo::planning::PlanningComponent::Proc(std::shared_ptrapollo::prediction::PredictionObstacles const&, std::shared_ptrapollo::canbus::Chassis const&, std::shared_ptrapollo::localization::LocalizationEstimate const&) ()
> #12 0x00000000004f9e19 in apollo::simulation::PlanningTask::RunOnce() ()
> #13 0x000000000048407b in apollo::simulation::TaskManager::ExecuteNextTaskAndWrite() ()
> #14 0x000000000046df39 in apollo::simulation::ScenarioRunner::Run() ()
> #15 0x00000000004351bb in main ()

This problem have already solved. The reason is "RuleBasedStopDecider::injector_" didn't assign a initial value.

**Comment by unacao** [comment]
@yifeijiang @xiaoxq @kechxu @Capri2014 simulation results for public and core scenarios are the same as current master branch. Please help review the code changes.

**Pull request merged**